connection-service/connection-flask.py

```diff
@@ -7,204 +7,244 @@
-      if ep in store['connection']:
-        store['connection'].pop(ep)
-        store['endpoint'].pop(ep)
-        partitions[part]=store
-        cmgr.update(partitions)
-        return 'OK'
-      else:
-        abort(404)
+        print(f"retract() could not find connection_id <{id}>")
+        good=False
+    if len(store)==0:
+      # We've deleted the last entry in this partition so delete the
+      # partition as well
+      partitions.pop(part)
   else:
-    abort(404)
-
-@app.route("/getendpoint/<part>",methods=['POST','GET'])
-def get_endpoint(part):
-  if part in partitions:
-    store=partitions[part]
-
-    endpoint=json.loads(request.form['endpoint'])
-    matches=[]
-    ctype=""
-    for entry in store['endpoint']:
-      match=True
-      ep=json.loads(entry)
-      for item in endpoint:
-        #print(f"item=<{item}> entry=<{ep[item]}> search=<{endpoint[item]}>")
-        if ep[item]!=endpoint[item] and endpoint[item] != '*':
-          match=False
-      if match:
-        matches.append(store['endpoint'][entry])
-        if entry in store['connection']:
-          conn=store['connection'][entry]
-          ctype=conn['connection_type']
-        if entry in store['connection_type']:
-          ctype=store['connection_type'][entry]
-    if len(matches)>0:
-      result='{'+f'"uris":['
-      for n in range(len(matches)):
-        result=result+'"'+matches[n]+'"'
-        if n<len(matches)-1:
-          result=result+','
-      result=result+'],'+f'"connection_type":"{ctype}"'+'}'
-      #print(f"result = <{result}>")
-      dummy=json.loads(result)
-      print(f"json   = <{result}>")
-      return(result)
-    else:
-      abort(404)
+    good=False
+  partlock.release()
+  if good:
+    return 'OK'
   else:
-    #print(f"Partition {part} not found")
     abort(404)
 
-def lookup(part, map, key):
-  if part in partitions:
-    store=partitions[part]
-    #print(f"Looking for <{key}> in <{store[map]}>")
-    if key in store[map]:
-      val=store[map][key]
-      #print(f"lookup key={key} {map}[key]={val}")
-      return(val)
+@app.route("/getconnection/<part>",methods=['POST','GET'])
+def get_connection(part):
+  # Find connection uris that correspond to the connection id pattern
+  # in the request. The pattern is treated as a regular expression.
+  now=datetime.now()
+  js=json.loads(request.data)
+  if debug_level>2:
+    print (f"[{now}] get_connection() {js=}")
+
+  if 'uid_regex' in js and 'data_type' in js:
+    if debug_level>1:
+      print(f"[{now}] get_connection()"
+            f" Searching for connections matching uid_regex<{js['uid_regex']}>"
+            f" and data_type {js['data_type']}")
+    result=[]
+    regex=re.compile(js['uid_regex'])
+    dt=js['data_type']
+    partlock.acquire()
+    if part in partitions:
+      store=partitions[part]
+      matched=[]
+      for uid,con in store.items():
+        if regex.search(uid) and con.data_type==dt and now-con.time<entry_ttl:
```

Comment:
`con.data_type==dt` why do we need this? Are there cases where where the uid is the same but data type is different?

`now-con.time<entry_ttl` Just so that I understand, this is the timeout right?

Reply:
Well the uid is selected by regex so the expression could match anything and you may just get everything with a given data type.

Yes, ttl is the timeout.

Reply:
Gotcha, so you can query on the data type.